hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/space/SpaceEnricher.java

```diff
@@ -0,0 +1,90 @@
+package org.hypertrace.traceenricher.enrichment.enrichers.space;
+
+import static org.hypertrace.traceenricher.enrichedspan.constants.EnrichedSpanConstants.SPACE_IDS_ATTRIBUTE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.typesafe.config.Config;
+import io.grpc.Channel;
+import io.grpc.ManagedChannelBuilder;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.hypertrace.core.attribute.service.cachingclient.CachingAttributeClient;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.core.datamodel.shared.trace.AttributeValueCreator;
+import org.hypertrace.entity.data.service.client.EntityDataServiceClientProvider;
+import org.hypertrace.trace.reader.attributes.TraceAttributeReader;
+import org.hypertrace.traceenricher.enrichedspan.constants.utils.EnrichedSpanUtils;
+import org.hypertrace.traceenricher.enrichment.AbstractTraceEnricher;
+
+public class SpaceEnricher extends AbstractTraceEnricher {
+  private SpaceRulesCachingClient ruleClient;
+  private SpaceRuleEvaluator ruleEvaluator;
+
+  @Override
+  public void init(Config enricherConfig, EntityDataServiceClientProvider provider) {
+    super.init(enricherConfig, provider);
+
+    Channel configChannel =
+        ManagedChannelBuilder.forAddress(
+                enricherConfig.getString("config.service.config.host"),
+                enricherConfig.getInt("config.service.config.port"))
+            .usePlaintext()
+            .build();
+
+    Channel attributeChannel =
+        ManagedChannelBuilder.forAddress(
+                enricherConfig.getString("attribute.service.config.host"),
+                enricherConfig.getInt("attribute.service.config.port"))
+            .usePlaintext()
+            .build();
+
+    // TODO - we need a way to share caching clients like the attribute reader across enrichers
+    this.init(
+        new SpaceRulesCachingClient(configChannel),
+        new SpaceRuleEvaluator(
+            TraceAttributeReader.build(CachingAttributeClient.builder(attributeChannel).build())));
+  }
+
+  /**
+   * The current design of enrichers with required no arg constructors does not allow mocking
+   * dependencies, so exposing an init method to support mocking clients for tests
+   */
+  @VisibleForTesting
+  void init(SpaceRulesCachingClient ruleClient, SpaceRuleEvaluator ruleEvaluator) {
+    this.ruleClient = ruleClient;
+    this.ruleEvaluator = ruleEvaluator;
+  }
+
+  @Override
+  public void enrichEvent(StructuredTrace trace, Event event) {
+    addEnrichedAttribute(
```

Comment:
Actually, I think we can do optimization by running this method only once on the head span of an API boundary and copying that to the rest of the spans. Take a look at the usage of ApiNode class. A structure trace is nothing but a list of ApiNodes.

Reply:
With the exception of the cached remote calls (which are unavoidable in the current design, but low frequency due to a very high cache hit rate), the rest of the path is constant time. It's a bit fancy in the implementation since its dynamic and leaves room for just in time fetches and other rule types, but all it's doing is reading an attribute value. 

IMO, it's a premature optimization to try to reuse the calculated value because: 
1. It gains little in the way of performance, if anything. Building the graph will actually be significantly more expensive (looks like we use static thread locals to try to share this cost though?), and using it will be more complex but should be comparable in performance.
2.  It puts some strong assumptions on how space values will be generated. The use case around environments should behave the way you described (sharing one value across a given api boundary), but spaces in general need not.